lib/multipart_form.ml

```diff
@@ -199,141 +189,156 @@ module QP = struct
-        let chunk' = Bytes.create (String.length chunk + 1) in
-        Bytes.blit_string chunk 0 chunk' 0 (String.length chunk) ;
-        check_end_of_body >>= choose chunk'
+          (* [pecu] expects inputs. At this stage, we know that after [chunk], we
+             have the first byte of [end_of_body] - but we don't know if we have
+             [end_of_body] or a part of it.
+
+             [check_end_of_body] will advance to see if we really have
+             [end_of_body]. The result will be sended to [choose]. *)
+          let chunk' = Bytes.create (String.length chunk + 1) in
+          Bytes.blit_string chunk 0 chunk' 0 (String.length chunk) ;
+          check_end_of_body >>= choose chunk'
       | `Data data ->
-        write_data data ; go ()
+          write_data data ;
+          go ()
       | `Line line ->
-        write_line line ; go ()
+          write_line line ;
+          go ()
       | `Malformed err -> fail err in
     go ()
 
-  let with_push ?(end_of_line = "\n") ~push end_of_body =
-    let write_data x = push (Some x) in
-    let write_line x =
-      push (Some (x ^ end_of_line)) in
+  let with_emitter ?(end_of_line = "\n") ~emitter end_of_body =
+    let write_data x = emitter (Some x) in
+    let write_line x = emitter (Some (x ^ end_of_line)) in
     parser ~write_data ~write_line end_of_body
 
   let to_end_of_input ~write_data ~write_line =
     let dec = Pecu.decoder `Manual in
 
-    fix @@ fun m -> match Pecu.decode dec with
+    fix @@ fun m ->
+    match Pecu.decode dec with
     | `End -> commit
-    | `Await ->
-      (peek_char >>= function
-        | None -> Pecu.src dec Bytes.empty 0 0 ; return ()
-        | Some _ -> available >>= fun n -> Unsafe.take n
-            (fun ba ~off ~len ->
-              let chunk = Bytes.create len in
-              Bigstringaf.blit_to_bytes ba ~src_off:off chunk ~dst_off:0 ~len ;
-              Pecu.src dec chunk 0 len)
-          >>= fun () -> m)
-    | `Data data -> write_data data ; m
-    | `Line line -> write_line line ; m
+    | `Await -> (
+        peek_char >>= function
+        | None ->
+            Pecu.src dec Bytes.empty 0 0 ;
+            return ()
+        | Some _ ->
+            available >>= fun n ->
+            Unsafe.take n (fun ba ~off ~len ->
+                let chunk = Bytes.create len in
+                Bigstringaf.blit_to_bytes ba ~src_off:off chunk ~dst_off:0 ~len ;
+                Pecu.src dec chunk 0 len)
+            >>= fun () -> m)
+    | `Data data ->
+        write_data data ;
+        m
+    | `Line line ->
+        write_line line ;
+        m
     | `Malformed err -> fail err
-
-  let to_end_of_input_with_push ?(end_of_line = "\n") push =
-    let write_data x = push (Some x) in
-    let write_line x = push (Some (x ^end_of_line)) in
-    to_end_of_input ~write_data ~write_line
 end
 
+type 'a elt = { header : Header.t; body : 'a }
+
+type 'a t = Leaf of 'a elt | Multipart of 'a t option list elt
```

Comment:
this signature looks a little different than before -- was that intentional? Feels like there's more nesting

Reply:
Yes, I think the type is less complex and of course, it did not highlight real cases, but:
1) a part can contain an other multipart (`'a t`)
2) a part can be empty (`option`)
3) a multipart is a list of parts (`list`)
4) and a multipart contains a header (`elt`)

When I implemented tests, I discovered n useful function such as `'a t -> (string * 'a) list` (where the `string` is the name of the field - given by `Content-Disposition`) and facilitate the processing of a `multipart-form` for the end-user. I will try to improve the API next week.